tempesta_fw/sock_srv.c

```diff
@@ -213,28 +218,29 @@ static int
 tfw_sock_srv_connect_failover(struct sock *sk)
 {
 	int r;
-	TfwSrvConnection *srv_conn = sk->sk_user_data;
-	TfwServer *srv = (TfwServer *)srv_conn->conn.peer;
+	TfwConnection *conn = sk->sk_user_data;
+	TfwServer *srv = (TfwServer *)conn->peer;
 
 	TFW_DBG_ADDR("connection lost", &srv->addr);
 
-	/* Revert tfw_sock_srv_connect_complete(). */
-	tfw_sg_update(srv->sg);
-	tfw_connection_destruct(&srv_conn->conn);
+	/* Withdraw from socket activity. */
+	spin_lock(&conn->splock);
+	tfw_connection_unlink_sk(conn, sk);
+	spin_unlock(&conn->splock);
```

Comment:
Also it seems there is nothing to synchronize. There are other similar locking places at the below, need to carefully review the locks. If the locks are needed, then probably it's better to move them to tfw_connection_unlink_sk() and similar functions.